File: solution/1356____Sort Integers by The Number of 1 Bits.go
package main

import "sort"

func sortByBits(arr []int) []int {
	return o1356(arr)
}

type o1356_struct struct {
	nums []int
	bits []int
}

func (o *o1356_struct) Len() int {
	return len(o.nums)
}

func (o *o1356_struct) Less(i, j int) bool {
	if o.bits[i] < o.bits[j] {
		return true
	} else if o.bits[i] > o.bits[j] {
		return false
	} else {
		return o.nums[i] < o.nums[j]
	}
}

func (o *o1356_struct) Swap(i, j int) {
	o.nums[i], o.nums[j] = o.nums[j], o.nums[i]
	o.bits[i], o.bits[j] = o.bits[j], o.bits[i]
}

// perf: 72, 56
func o1356(arr []int) []int {
	bits := make([]int, len(arr), len(arr))
	o1356_count_bits(arr, bits)
	temp := o1356_struct{arr, bits}
	sort.Sort(&temp)
	return temp.nums
}

func o1356_count_bits(arr, bits []int) {
	for mask := 1; mask != 0; mask <<= 1 {
		for i, v := range arr {
			if v&mask != 0 {
				bits[i]++
			}
		}
	}
}

func main() {

}

File: solution/1356____Sort_Integers_by_The_Number_of_1_Bits.cpp

#include "util\frequently_included.h"

int bit_count(int i){
        int ans = 0;
        for(unsigned mask = 1; mask; mask <<= 1)
                if(mask & i)
                        ++ans;
        return ans;
}

bool cmp(const int& i, const int& ii){
        int diff = bit_count(i) - bit_count(ii);
        if(diff < 0)
                return true;
        if(diff > 0)
                return false;
        return i < ii;
}

class Solution {
public:
        vector<int> sortByBits(vector<int>& arr) {
                _1(arr);
                return arr;
        }
private:
        void _1(vector<int>& arr){
                sort(arr.begin(), arr.end(), cmp);
        }
};

int main(){
        vector<int> arr = {0,1,2,3,4,5,6,7,8};
        Solution sln;
        sln.sortByBits(arr);
        return 0;
}

